### PR TITLE
[lld-macho] Implement support for ObjC relative method lists

### LINKED ISSUE #57809: ld64.lld: Support synthesizing relative ObjC method lists
Starting with macOS 11, a more compact encoding is available for ObjC method lists. Previously, each method had a descriptor consisting of 3 pointers:
```c
struct _objc_method {
    SEL _cmd;
    char *method_type;
    char *_imp;
}
```

In the new format, `__TEXT,__objc_methlist` stores 32-bit offsets instead.

```c++
struct _objc_method_rel {
    int32_t _cmd;
    int32_t method_type;
    int32_t _imp;
}
```

This transformation can be controlled with the `-objc_relative_method_lists`/`-no_objc_relative_method_lists` linker flags.

Currently, method lists are stored in `__DATA,__objc_const`. With this transformation, Chromium Framework's data segment will shrink by about 40 KB, meaning that 3 fewer pages will have to be relocated at startup.

ref #48630

<details>

<summary>Sidenote: AppleClang itself can be coaxed into generating the relative encoding with `-fobjc-relative-method-lists`</summary>

```objc
// test.m
#include <Foundation/Foundation.h>

int main() {
    NSObject *foo = [[NSObject alloc] init];
}
```

```console
❯ clang test.m  -S -emit-llvm  -o - | grep _objc_method
%struct.__method_list_t = type { i32, i32, [0 x %struct._objc_method] }
%struct._objc_method = type { i8*, i8*, i8* }

❯ clang test.m  -S -emit-llvm  -o - -fobjc-relative-method-lists | grep _objc_method
%struct.__method_list_t = type { i32, i32, [0 x %struct._objc_method_rel] }
%struct._objc_method_rel = type { i32, i32, i32 }
```
</details>

### PR DESCRIPTION
**Pull request by alx32**
The MachO format supports relative offsets for ObjC method lists. This support is present already in ld64. With this change we implement this support in lld also.

Relative method lists can be identified by a specific flag (0x80000000) in the method list header. When this flag is present, the method list will contain 32-bit relative offsets to the current Program Counter (PC), instead of absolute pointers.
Additionally, when relative method lists are used, the offset to the selector name will now be relative and point to the selector reference (selref) instead of the name itself.

### CODE REVIEW

File: lld/MachO/SyntheticSections.cpp
Comment (kyulee-com) [inline lld/MachO/SyntheticSections.cpp:2125]: The same here on `sizeof(uint32_t)`. Is it `align`?

File: lld/test/MachO/objc-relative-method-lists-simple.s
Comment (kyulee-com) [inline lld/test/MachO/objc-relative-method-lists-simple.s:8]: Would you also use `%no-arg-lld` or `%lld`?

File: lld/test/MachO/objc-relative-method-lists-simple.s
Comment (kyulee-com) [inline lld/test/MachO/objc-relative-method-lists-simple.s:15]: I don't see other checks with `CHK_NO_REL` other than this line. Can we check the original output of method lists?

**Review by kyulee-com — COMMENTED** [review]
Thank you for working on this! Overall it looks good.

File: lld/MachO/Driver.cpp
Comment (alx32) [reply lld/MachO/Driver.cpp:1094]: Maybe I'm misunderstanding, but doesn't `args.getLastArg` get the last command line arg that matches. If both are passed, then only the last one is considered. Does the order matter here ?

File: lld/test/MachO/objc-relative-method-lists-simple.s
Comment (kyulee-com) [inline lld/test/MachO/objc-relative-method-lists-simple.s:52]: Not sure if addresses are constant, but you might want to relax it using a regex like the case with relative method list.

**Review by kyulee-com — APPROVED** [review]
lgtm modulo minor comment on test.

**Comment by BertalanD** [comment]
Feel free to mark this as closing #57809

**Comment by kyulee-com** [comment]
thanks for fixing the header in place before merge!

**Comment by luporl** [comment]
I have disabled the objc-relative-method-lists-simple.s test on 32-bit ARM targets, in 6b95747087032e8f86af64325780d8471e6fce58, to fix failures with this buildbot:
https://lab.llvm.org/buildbot/#/builders/178/builds/7227

**Comment by alx32** [comment]
Thanks for the fix! Is there any way to run these tests before merging a change ?
Is there a guide on what hardware to get to run these tests on ?

**Comment by luporl** [comment]
> Thanks for the fix! Is there any way to run these tests before merging a change ? Is there a guide on what hardware to get to run these tests on ?

I'm afraid there is no easy way to run these tests before merging a change. Pre-commit CI covers the more common cases.
For this specific test you would need an ARMv8 machine (or VM) running Linux with armhf userspace.
(the performance of running builders directly on armhf hardware is not good)

LLVM post-commit buildbots can be found here: https://lab.llvm.org/buildbot.
The subset of bots maintained by Linaro, that are focused on ARM, can be more easily checked here: http://llvm.validation.linaro.org.

**Pull request merged**
